### ISSUE TITLE
Add Offline DeepSeek Model

### ISSUE THREAD

**Issue by Barqawiz** [intelligentnode/Intelli#82]
Implement an offline DeepSeek model loader for inference that:

- Loads DeepSeek models directly from the official host (HuggingFace).
- Supports both full and quantized versions (if available).
- Implement memory optimization techniques similar to llama.cpp or ollama.

**Expected Deliverables**:

- Develop a [wrapper](https://github.com/intelligentnode/Intelli/tree/main/intelli/wrappers) for the DeepSeek model.
- Create a folder under [model](https://github.com/intelligentnode/Intelli/tree/main/intelli/model)/deepseek for helpers and extended functions used by the wrapper.
- Write a test case to load the model, to be placed in the [test](https://github.com/intelligentnode/Intelli/tree/main/intelli/test/integration) directory.
- Mention which pip used for the code like torch. Don't use higher level modules like transformers.

**Notes**:
- For reference, check the current (non-optimized) Python code from DeepSeek repo: 
https://github.com/deepseek-ai/DeepSeek-V3/tree/main/inference
- llama.cpp is a reference for optimized model loading techniques.
- Intelli should provide easy way to load the model.
- **Don't use** high level pip module like transformers (intellinode provide light weight integration for AI agents).
- You can use torch, tensorflow, keras, safetensors, triton, etc. (**these modules recommended**)

**Comment by oliverqx** [comment]
@Barqawiz  could you share the reasoning behind opting for a custom model loader? Have you considered leveraging Ollama for loading and running the DeepSeek models instead?

By building a lightweight wrapper to integrate with Ollama, we could create a unified API for AI agents that not only supports DeepSeek but also any other models compatible with Ollama. This would simplify development, reduce overhead, and ensure we're tapping into the optimizations that Ollama offers for efficient model inference and memory management.

This approach would make it easier to scale and support a broader range of models in the future

**Comment by intelligentnode** [comment]
Good question. Let me explain the rationale behind using an offline model and why I'm avoiding Ollama or similar high-level modules.

Intelli can build a graph of collaborating agents using the flow concept: [Sequence Flow Documentation](https://docs.intellinode.ai/docs/python/flows/sequence-flow).

I've managed to integrate multiple offline models into the flow using the KerasWrapper, which provides a convenient way to load several offline models, such as Llama, Mistral, and others: [KerasWrapper class](https://github.com/intelligentnode/Intelli/blob/main/intelli/wrappers/keras_wrapper.py).

However, Keras does not currently support DeepSeek, and adding that functionality will likely take some time. As a result, my current focus is on DeepSeek only.

I avoid using Ollama because I want to minimize external dependencies. I looked into Ollama as a high-level library, and integrating it would introduce additional unnecessary modules - the same thing with Transformers. 

You can influence how  Ollama uses modules like Torch, optimization libraries, or Safetensors from HuggingFace. These lower-level modules are accepted. I'm happy to credit their work if you influence their approaches, but I prefer not to have Ollama as a dependency for running the flow.

**Comment by oliverqx** [comment]
@intelligentnode are there any specific DeepSeek variants you'd prefer?

**Comment by intelligentnode** [comment]
You can use the official ones from **R1 or any quantized variant**:

**DeepSeek-R1 Models**

<div align="center">

| **Model** | **#Total Params** | **#Activated Params** | **Context Length** | **Download** |
| :------------: | :------------: | :------------: | :------------: | :------------: |
| DeepSeek-R1   | 671B | 37B |  128K   | [🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-R1)   |

</div>

**DeepSeek-R1-Distill Models**

<div align="center">

| **Model** | **Base Model** | **Download** |
| :------------: | :------------: | :------------: |
| DeepSeek-R1-Distill-Qwen-1.5B  | [Qwen2.5-Math-1.5B](https://huggingface.co/Qwen/Qwen2.5-Math-1.5B) | [🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B)   |
| DeepSeek-R1-Distill-Qwen-7B  | [Qwen2.5-Math-7B](https://huggingface.co/Qwen/Qwen2.5-Math-7B) | [🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Qwen-7B)   |
| DeepSeek-R1-Distill-Llama-8B  | [Llama-3.1-8B](https://huggingface.co/meta-llama/Llama-3.1-8B) | [🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Llama-8B)   |
| DeepSeek-R1-Distill-Qwen-14B   | [Qwen2.5-14B](https://huggingface.co/Qwen/Qwen2.5-14B) | [🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Qwen-14B)   |
|DeepSeek-R1-Distill-Qwen-32B  | [Qwen2.5-32B](https://huggingface.co/Qwen/Qwen2.5-32B) | [🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Qwen-32B)   |
| DeepSeek-R1-Distill-Llama-70B  | [Llama-3.3-70B-Instruct](https://huggingface.co/meta-llama/Llama-3.3-70B-Instruct) | [🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Llama-70B)   |
---

In general it is going to be expensive to run DeepSeek-R1.  But you can test the code on the 7B models to be accept as attempt. Also if you know a hosted quantized version from R1 you can test on it.

**Comment by intelligentnode** [comment]
If it requires installing Transformers to use it, then no.
If it is published as an independent module with a lightweight installation via pip, then yes.
You can implement a lightweight version of Tokenizer, if an independent one is not available.

**Comment by oliverqx** [comment]
@Enity300 would you like to work on this together? its quite the task i think

**Comment by intelligentnode** [comment]
@oliverqx It is good that you mentioned this. With your collaboration, I can assist you with relevant chunks of the code that require stitching, and testing. 

Send me your email using the below form, and will organize a call:
https://www.intellinode.ai/contact

Mention you are from Github.

**Comment by oliverqx** [comment]
@intelligentnode 

so far ive studied the deep seek repo, this week ive been studying llama.cpp, sometime this weekend i think i can open a PR. 

until then, work is very theoretical. Will definitly hit you up once i get a solid grasp on optimization

**Comment by RaghavArora14** [comment]
Hi @intelligentnode I recently submitted a PR implementing this bounty on behalf of @Enity300 please review that too.

**Comment by Kunal-Darekar** [comment]
Hi @intelligentnode I recently submitted a PR implementing this bounty  please review that too.

**Comment by intelligentnode** [comment]
This task reward is time-sensitive, due to not having a complete solution until this time I have to adjust the /reward 300

**Comment by intelligentnode** [comment]
/reward 300

This task reward is time-sensitive, due to not having a complete solution until this time I have to adjust the reward.

**Comment by intelligentnode** [comment]
Cancelling as this issue is time-sensitive and not delivered on time.
